src/spell.c

```diff
@@ -160,72 +158,105 @@ char ilet;
 
 /* TRUE: book should be destroyed by caller */
 STATIC_OVL boolean
-cursed_book(bp)
-	struct obj *bp;
-{
+cursed_book(struct obj *bp){
 	int lev = objects[bp->otyp].oc_level;
+
+	boolean was_in_use;
+	boolean already_cursed = bp->cursed;
+	int dmg = 0;
+	
+	if(RoSbook == STUDY_WARD) lev -= 1; //The wardings partially protect you from the magic contained within the book.
 	
-	if(RoSbook == STUDY_WARD) lev -= 1; //The wardings paritally protect you from the magic contained within the book.
+	for (int spell = 0; spell < MAXSPELL; spell++)
+		if (spellid(spell) == bp->otyp)
+			lev -= 1; // if you already know it, you're more resilient to failure
+
+	if (lev < 0) lev = 0;
 	
 	switch(rn2(lev)) {
-	case 0:
-		You_feel("a wrenching sensation.");
-		tele();		/* teleport him */
-		break;
-	case 1:
-		You_feel("threatened.");
-		aggravate();
-		break;
-	case 2:
-		make_blinded(Blinded + rn1(100,250),TRUE);
-		break;
-	case 3:
-		take_gold();
-		break;
-	case 4:
-		pline("These runes were just too much to comprehend.");
-		make_confused(HConfusion + rn1(7,16),FALSE);
-		break;
-	case 5:
-		pline_The("book was coated with contact poison!");
-		if (uarmg) {
-		    if (uarmg->oerodeproof || !is_corrodeable(uarmg)) {
-			Your("gloves seem unaffected.");
-		    } else if (uarmg->oeroded2 < MAX_ERODE) {
-			if (uarmg->greased) {
-			    grease_protect(uarmg, "gloves", &youmonst);
+		case 0:
+			if (already_cursed || rn2(4)) {
+				You_feel("a wrenching sensation.");
+				tele(); /* teleport him */
+			}
+			else {
+				if (bp->blessed) {
+					pline_The("book glows brown.");
```

Comment:
I believe it's technically possible to read a spellbook while hallucinating, so this could instead call hcolor:
`pline_The("book glows %s.", hcolor("brown"));`
And similarly for glowing black.